### ISSUE TITLE
MiniTest 5.11.0 -- error or failure also causes undefined method (for test name) backtrace

### ISSUE THREAD

**Issue by ccasabona** [seattlerb/minitest#730]
Minitest 5.11.0

When an error code E or a faulure code F is returned for a test, an 'undefined method for (test name here)' error with backtrace is generated.

Example:

```ruby
 test 'with an error' do
    assrt_equal 1, -1
 end 
```

The line fails because of assrt_equal with this backtrace:

```
Running via Spring preloader in process 3220
Run options: --seed 45600

# Running:

E

Error:
ShippingLabelIntegrationTest#test_with_an_error:
NoMethodError: undefined method `assrt_equal' for #<ShippingLabelIntegrationTest:0x00007fd8353a1c30>
Did you mean?  assert_equal
    test/integration/shipping_label_integration_test.rb:1013:in `block in <class:ShippingLabelIntegrationTest>'


/Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/railties-5.1.4/lib/rails/test_unit/reporter.rb:70:in `method': undefined method `test_with_an_error' for class `Minitest::Result' (NameError)
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/railties-5.1.4/lib/rails/test_unit/reporter.rb:70:in `format_rerun_snippet'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/railties-5.1.4/lib/rails/test_unit/reporter.rb:23:in `record'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:791:in `block in record'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:790:in `each'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:790:in `record'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:334:in `run_one_method'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:321:in `block (2 levels) in run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:320:in `each'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:320:in `block in run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:360:in `on_signal'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:347:in `with_info_handler'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:319:in `run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/railties-5.1.4/lib/rails/test_unit/line_filtering.rb:9:in `run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:159:in `block in __run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:159:in `map'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:159:in `__run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:136:in `run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:63:in `block in autorun'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/spring-2.0.2/lib/spring/application.rb:171:in `fork'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/spring-2.0.2/lib/spring/application.rb:171:in `serve'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/spring-2.0.2/lib/spring/application.rb:141:in `block in run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/spring-2.0.2/lib/spring/application.rb:135:in `loop'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/spring-2.0.2/lib/spring/application.rb:135:in `run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/spring-2.0.2/lib/spring/application/boot.rb:19:in `<top (required)>'
	from /Users/charlie/.rvm/rubies/ruby-2.4.3/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	from /Users/charlie/.rvm/rubies/ruby-2.4.3/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	from -e:1:in `<main>'
`
 When corrected and the test is run:

 test 'with an error' do
    assert_equal 1, -1
 end

This time the test fails with a backtrace:

`Running via Spring preloader in process 3290
Run options: --seed 27529

# Running:

F

Failure:
ShippingLabelIntegrationTest#test_with_an_error [/Users/charlie/sites/defluris_engine/defluris/test/integration/shipping_label_integration_test.rb:1013]:
Expected: 1
  Actual: -1


/Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/railties-5.1.4/lib/rails/test_unit/reporter.rb:70:in `method': undefined method `test_with_an_error' for class `Minitest::Result' (NameError)
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/railties-5.1.4/lib/rails/test_unit/reporter.rb:70:in `format_rerun_snippet'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/railties-5.1.4/lib/rails/test_unit/reporter.rb:23:in `record'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:791:in `block in record'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:790:in `each'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:790:in `record'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:334:in `run_one_method'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:321:in `block (2 levels) in run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:320:in `each'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:320:in `block in run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:360:in `on_signal'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:347:in `with_info_handler'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:319:in `run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/railties-5.1.4/lib/rails/test_unit/line_filtering.rb:9:in `run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:159:in `block in __run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:159:in `map'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:159:in `__run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:136:in `run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/minitest-5.11.0/lib/minitest.rb:63:in `block in autorun'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/spring-2.0.2/lib/spring/application.rb:171:in `fork'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/spring-2.0.2/lib/spring/application.rb:171:in `serve'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/spring-2.0.2/lib/spring/application.rb:141:in `block in run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/spring-2.0.2/lib/spring/application.rb:135:in `loop'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/spring-2.0.2/lib/spring/application.rb:135:in `run'
	from /Users/charlie/.rvm/gems/ruby-2.4.3@ruby243rails51/gems/spring-2.0.2/lib/spring/application/boot.rb:19:in `<top (required)>'
	from /Users/charlie/.rvm/rubies/ruby-2.4.3/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	from /Users/charlie/.rvm/rubies/ruby-2.4.3/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:59:in `require'
	from -e:1:in `<main>'
```

Under Minitest 5.10.3, the no method error and backtrace do not appear.
 
* Rails 5.1.4
* Ruby 2.4.3
* Minitest 5.11.0

**Comment by ChrisBr** [comment]
Same here, seems to be a problem with Rails::TestUnitReporter

```
def format_rerun_snippet(result)
        location, line = result.method(result.name).source_location
        "#{executable} #{relative_path_for(location)}:#{line}"
end
```

As Minitest passes now a Minitest::Result object to the reporter which does not have the test method:
https://github.com/seattlerb/minitest/commit/00433fc0a4fdd0e6b302aace633384ba13122376#diff-efcebb0b0e7d5c021792f400fd891d9fR449

**Comment by ChrisBr** [comment]
Storing the runnable in the result objet and adding an around alias for the ``method`` method fixes the issue for me. However, I don't really like this solution. @zenspider any better idea?

```rb
    def self.from runnable
      o = runnable

      r = self.new o.name
      r.klass      = o.class.name
      r.assertions = o.assertions
      r.failures   = o.failures.dup
      r.time       = o.time
      r.runnable = o

      r.source_location = o.method(o.name).source_location rescue ["unknown", -1]

      r
    end

    alias :orig_method :method

    def method method_name
      self.orig_method(method_name)
    rescue
      self.runnable.method(method_name)
    end
```

**Comment by zenspider** [comment]
UGH. I even asked @tenderlove to poke at this before the release. :(

I cannot and will not store the runnable in the Result. That's the whole point of the Result object.

Can you try:

```ruby
def format_rerun_snippet(result)
  location, line = if result.respond_to?(:source_location) then
                     result.source_location
                   else
                     result.method(result.name).source_location
                   end
  "#{executable} #{relative_path_for(location)}:#{line}"
end
```

I could also see doing a patch that has a dummy method object returned.

**Comment by zenspider** [comment]
Possible:

```ruby
class Minitest::Result
  def method name
    o = Object.new
    def o.source_location
      ["unknown", -1]
    end
  end
end
```

**Comment by zenspider** [comment]
Nevermind... This should be related to #729. Rails just makes it a lot messier looking. I have a fix that will be out shortly.

**Comment by zenspider** [comment]
Released 5.11.1. This should be fixed. Thanks!

**Comment by ChrisBr** [comment]
c & p from #732 

Unfortunately I can not reopen #730 so I'm opening a new one. Release 5.11.1 did not fix the issue, I still get the same error (see #730).

From my POV #729 and #730 are unrelated.

class Minitest::Result
  def method name
    o = Object.new
    def o.source_location
      ["unknown", -1]
    end
  end
end
Your suggestion did fix the issue, however, wouldn't that be always return ["unknown", -1] for source location? Also it would return just an Object instead of a MethodObject.

What about only storing the MethodObject instead of only the source location? This also worked for me:

def method name
  if name.to_sym == method_object.name
    method_object
  else
    orig_method(name)
  end
end
Your other suggestions would require to patch ActiveSupport, no?

Thanks!

**Comment by zenspider** [comment]
> What about only storing the MethodObject instead of only the source location?

Storing off the method maintains a binding to the original object, which is exactly what we're trying to avoid. I could try to anticipate the source location asked for and store that off?

**Comment by zenspider** [comment]
Here are two working proposals, one for patching minitest and one for rails. I might want both, tho the minitest one feels gross.

```ruby
require "minitest/autorun"

if ENV["MT"] then
  class Minitest::Result
    def method name
      self.instance_of?(Minitest::Result) && name == self.name ? self : super
    end
  end
end

describe "Whatever" do
  it 'with an error' do
    assert_equal 1, -1
  end

  def format_rerun_snippet(result) # from rails
    location, line =
      if ENV["RAILS"] then
        if result.respond_to?(:source_location) then
          result.source_location
        else
          result.method(result.name).source_location
        end
      else
        result.method(result.name).source_location
      end
    "executable #{location}:#{line}"
  end

  def run(*a) # emulate rails
    result = super
    puts format_rerun_snippet result unless result.passed?
    result
  end
end
```

**Comment by zenspider** [comment]
Linking to https://github.com/rails/rails/pull/31624

**Comment by y-yagi** [comment]
Thanks for your proposal. 
I think it will be very useful until released Rails that patched. 
Even if that proposal is executed, should check the `source_location` on Rails?

**Comment by jirihradil** [comment]
The same for me. Since minitest 5.11, each failure or error in tests aborts the whole testing with class `Minitest::Result' (NameError).

**Comment by zenspider** [comment]
I’m working with @tenderlove to figure out the best way to go with this. Until then, feel free to steal off the `MT` hunk from my repro+fix above.

**Comment by zenspider** [comment]
This has been fixed in rails. Thanks everyone. Poke me if there are futher problems.

**Comment by jm3** [comment]
@zenspider i see that tenderlove merged [Rails PR #31624](https://github.com/rails/rails/pull/31624) to master, but just confirming: that's not in any Rails release yet, right? Is the simplest thing for now to just lock MiniTest version to 5.10.3 in our Gemfiles until a new Rails release drops?

**Comment by acousma82** [comment]
@zenspider How has it been fixed in Rails. Do you mean the workaround with the gemfile.lock is the fix or is there a new Rails release with the fix?

**Comment by zenspider** [comment]
@acousma82 see the PR link in the comment above yours.

No, a release isn't out yet. You can run on edge rails to get this fix. (cc/ @jm3)

I'm not hooked in enough to know when a rails release will happen. Comment on the rails PR if you want to poke at it.

**Comment by davidcelis** [comment]
I'm seeing a `NoMethodError` in a non-Rails project, even just a small, single-file project (not using bundler). This is using Minitest 5.11.1:

```ruby
require 'minitest/autorun'
require 'set'

class TestSet < Minitest::Test
  def test_empty
    set= Set.new
    assert_empty set

    set.add(0)
    assert_empty set
  end
end
```

And, running the test:

```
$ ruby set_test.rb
Run options: --seed 53089

# Running:

E

Error:
TestSet#test_empty:
NoMethodError: undefined method `asset_empty' for #<TestSet:0x00007f98169f11f8>
Did you mean?  assert_empty
               test_empty
    set_test.rb:10:in `test_empty'


Traceback (most recent call last):
	17: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.1/lib/minitest.rb:63:in `block in autorun'
	16: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.1/lib/minitest.rb:136:in `run'
	15: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.1/lib/minitest.rb:159:in `__run'
	14: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.1/lib/minitest.rb:159:in `map'
	13: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.1/lib/minitest.rb:159:in `block in __run'
	12: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.1/lib/minitest.rb:319:in `run'
	11: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.1/lib/minitest.rb:347:in `with_info_handler'
	10: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.1/lib/minitest.rb:360:in `on_signal'
	 9: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.1/lib/minitest.rb:320:in `block in run'
	 8: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.1/lib/minitest.rb:320:in `each'
	 7: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.1/lib/minitest.rb:321:in `block (2 levels) in run'
	 6: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.1/lib/minitest.rb:334:in `run_one_method'
	 5: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.1/lib/minitest.rb:785:in `record'
	 4: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.1/lib/minitest.rb:785:in `each'
	 3: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/minitest-5.11.1/lib/minitest.rb:786:in `block in record'
	 2: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/railties-5.1.4/lib/rails/test_unit/reporter.rb:23:in `record'
	 1: from /Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/railties-5.1.4/lib/rails/test_unit/reporter.rb:70:in `format_rerun_snippet'
/Users/davidcelis/.rbenv/versions/2.5.0/lib/ruby/gems/2.5.0/gems/railties-5.1.4/lib/rails/test_unit/reporter.rb:70:in `method': undefined method `test_empty' for class `Minitest::Result' (NameError)
```

What I can't seem to figure out is why Rails' TestUnit reporter is getting loaded when I'm only requiring `minitest/autorun`. I have the Rails gem installed, but why is it getting required?

**Comment by ChadMoran** [comment]
Looking at your error it looks like you're trying to call asset_empty, not assert_empty.

> NoMethodError: undefined method `asset_empty' for #<TestSet:0x00007f98169f11f8>

**Comment by davidcelis** [comment]
@ChadMoran Whoops, that's true; I did originally have a typo. That trace appears regardless though, so I'll update that comment to avoid any further confusion

**Comment by zenspider** [comment]
@davidcelis You're commenting on an issue that was closed 2 weeks ago. It looks like while you are running tests in a non-rails project, you have rails globally installed and it is being activated and intruding. If you use `bundle exec` or remove rails from your global install it should be fine. The code works for me in a clean environment. This is more of an issue with how rails does things, not minitest.

**Comment by davidcelis** [comment]
@zenspider I know the issue's closed; I commented because I was experiencing it too. I see now that `Minitest.load_plugins` goes through any gem in the gemset for `minitest/*_plugin.rb` files and requires them, so that explains why Rails' test runner was getting loaded even though I hadn't required Rails. That does make me think it's still a Minitest thing and not a Rails thing, but I don't really manage any Ruby projects without bundler. I do wonder if Minitest could avoid loading plugins from gems that haven't actually been required (whether via `Gem.loaded_specs` or something else), but it isn't a real issue for me outside of this one file!

**Comment by hcfairbanks** [comment]
Work Around

Same error (undefined method (for test name) ) was happening to me with a freshly created rails app. I am running through rbenv, no globally installed rails. I removed minitest from the gemlock file and added gem minitest '5.10.3' to the gem file and that did the trick. I tried a bunch of different versions and it seems to happen on the following versions
5.11.0b1
5.11.0
5.11.1 
5.11.2
5.11.3 
I know this is a closed issue but people having trouble and this is something that could help. 

Environment
kubuntu 16.04
rbenv 1.1.0-2-g4f8925a
minitest 5.10.3
rails 5.1.0 (also works for 5.14 and I imagine versions in between)
ruby 2.4.0p0 (2016-12-24 revision 57164) [x86_64-linux]

No Global rails gem
deploy@deploy-VirtualBox ~ $ gem list global
*** LOCAL GEMS ***
globalid (0.4.1, 0.4.0)
deploy@deploy-VirtualBox ~ $
